File: android/app/src/main/java/com/noop/managed/ManagedAuthenticationRetry.kt
package com.noop.managed

/**
 * Replays one idempotent managed-storage operation after refreshing both Firebase credentials.
 *
 * The operation owns its durable checkpoints. Only authentication failures are retried, and the
 * refreshed attempt is never looped.
 */
internal object ManagedAuthenticationRetry {
    suspend fun <T> run(
        authorization: suspend (forceRefresh: Boolean) -> ManagedAuthorization,
        operation: suspend (authorization: ManagedAuthorization) -> T,
    ): T = try {
        operation(authorization(false))
    } catch (_: ManagedStorageException.Authentication) {
        operation(authorization(true))
    }
}

File: Packages/NoopRemoteSync/Sources/NoopRemoteSync/ManagedAuthenticationRetry.swift
public enum ManagedAuthenticationRetry {
    /// Replays one idempotent managed-storage operation after refreshing both Firebase credentials.
    ///
    /// The caller owns the operation's durable checkpoints. This helper deliberately retries only
    /// authentication failures and only once; policy, quota, integrity, and transport failures retain
    /// their original behavior.
    public static func run<Result>(
        authorization: (_ forceRefresh: Bool) async throws -> ManagedAuthorization,
        operation: (_ authorization: ManagedAuthorization) async throws -> Result
    ) async throws -> Result {
        do {
            return try await operation(authorization(false))
        } catch ManagedStorageError.authentication {
            return try await operation(authorization(true))
        }
    }
}
